Use general word instead of "Concurrency" in order to support other metric in KPA

yanweiguo opened the pull request:
<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Part of  #3416

We want to support other metric other than concurrency in KPA. In order to do so, rename some fields, variables, functions to be more general.

## Proposed Changes

* Rename fileds: `Decider.Spec.TargetConcurrency` to `Decider.Spec.TargetValue`, `Decider.Spec.TotalConcurrency` to `Decider.Spec.TotalValue`.
* Rename variables: `observedStableConcurrency` to `observedStableValue`, `observedPanicConcurrency` to `observedPanicValue`
* Rename function: `ResolveConcurrency` to `ResolveMetricTarget`.

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
NONE
```

vagababov commented on pkg/autoscaler/autoscaler.go:
```diff
@@ -139,21 +139,21 @@ func (a *Autoscaler) Scale(ctx context.Context, now time.Time) (desiredPodCount
+	a.reporter.ReportPanicRequestConcurrency(observedPanicValue)
+	a.reporter.ReportTargetRequestConcurrency(spec.TargetValue)
 
 	logger.Debugw(fmt.Sprintf("Observed average %0.3f concurrency, targeting %0.3f.",
```
some still left

yanweiguo replied:
Rename some more. There're some left. I'll change them when we know which metric is used. Currently no way to tell and concurrency is the only one supported.

yanweiguo commented:
> I am fine in general with the approach, but I'd let @markusthoemmes take a look as well, since it's still a bit unclear whether we're using the same `target` annotation (which is fine) and the default/containerConcurrency values (which sound wrong for this).

For `target` annotations, currently I don't see any whose name is restricted to concurrency. For some configurations in configmap or Spec with concurrency in the name, I don't think we want to reuse them for other metrics since it will cause confusions.

markusthoemmes commented on pkg/reconciler/autoscaling/resources/target.go:
```diff
@@ -23,11 +23,11 @@ import (
+// `target` is the target value of scaling metric that we autoscaler will aim for;
+// `total` is the maximum possible value of scaling metric that is permitted on the pod.
+func ResolveMetricTarget(pa *v1alpha1.PodAutoscaler, config *autoscaler.Config) (target float64, total float64) {
 	total = float64(pa.Spec.ContainerConcurrency)
```
As Victor mentioned, this feels wrong now. Does it need to be sensitive to the Target annotation to pick the right defaults from config?

yanweiguo replied:
Yes it needs. However currently we only support concurrency.
I didn't change the logic here or introducing new fields since this PR is just about renaming.

What we currently have for scaling metrics and targets are:
1. Annotations:
```
autoscaling.knative.dev/target
autoscaling.knative.dev/targetUtilizationPercentage
autoscaling.knative.dev/targetBurstCapacity
```

They can be reused for other metrics.

2. Configurations in config-autoscaler:
```
target-burst-capacity
container-concurrency-target-percentage
container-concurrency-target-default
```

The first one can be reused for other metric. The later two can not.

3. Spec in Revision:
```
Spec.ContainerConcurrency
```

This is concurrency only.


My plan is to reuse all the three existing annotations and introduce two more field in configmap:
```
target-percentage
requests-per-second-target-default
```
Use these 5 knobs to calculate target value of RPS(BTW I want to rename OPS to RPS as well). So there will be logic change here after the new fields are introduced.

Does it look good to you?

yanweiguo replied:
Added a TODO as well.

yanweiguo commented:
@markusthoemmes friendly ping

markusthoemmes commented on pkg/autoscaler/multiscaler.go:
```diff
@@ -65,7 +65,7 @@ type DeciderStatus struct {
 
 	// ExcessBurstCapacity is the difference between spare capacity
-	// (how many more concurrent requests the pods in the revision
+	// (how many more tokens in terms of scaling metrics the pods in the revision
```
maybe we should generalize this a bit differently? Something like

```
(how much more load the pods in the revision deployment can take before being overloaded)
```

yanweiguo replied:
Done.

yanweiguo commented:
/retest

yanweiguo commented:
/retest

yanweiguo commented:
/test pull-knative-serving-unit-tests
/test pull-knative-serving-build-tests

markusthoemmes commented:
Retries won't solve the build issue:

```
 # knative.dev/serving/pkg/reconciler/autoscaling/kpa [knative.dev/serving/pkg/reconciler/autoscaling/kpa.test]
pkg/reconciler/autoscaling/kpa/kpa_test.go:1563:4: unknown field 'TargetConcurrency' in struct literal of type autoscaler.DeciderSpec
pkg/reconciler/autoscaling/kpa/kpa_test.go:1564:4: unknown field 'TotalConcurrency' in struct literal of type autoscaler.DeciderSpec 
```

yanweiguo commented:
>  # knative.dev/serving/pkg/reconciler/autoscaling/kpa [knative.dev/serving/pkg/reconciler/autoscaling/kpa.test]
> pkg/reconciler/autoscaling/kpa/kpa_test.go

Interesting. The tests passed before I changed some comment so I think the failure was not due to my change.

Merged the master and renamed the remained.

markusthoemmes commented on pkg/autoscaler/multiscaler.go:
```diff
@@ -65,8 +65,8 @@ type DeciderStatus struct {
-	// (how many more concurrent requests the pods in the revision
-	// deployment can serve) and the configured target burst capacity.
+	// (how much more load the pods in the revision deployment can take before being
+    // overloaded) and the configured target burst capacity.
```
fix the indentation please.

yanweiguo replied:
ah, table vs spaces. Done.
